### PR TITLE
Speed up host padding

### PR DESCRIPTION
**Pull request by philei-tt**
### Ticket
N/A

### Problem description
When loading weights and inputs to the model, we usually call `ttnn.from_torch`, which in case of ttnn.TILE_LAYOUT and shape not aligned by ttnn.TILE_SIZE will call padding on host before moving data to device.  Current host padding is non-optimal. Speeding it up will improve both model start time and overall model execution time.  

### What's changed
- Rewrite padding with explicit index calculation, std::memcpy the whole row and no recursion
- Add c++ padding benchmark to verify padding speeding up ([tests/tt_metal/tt_metal/benchmark/padding/pad_rm.cpp](https://github.com/tenstorrent/tt-metal/blob/philei/speed-up-host-padding/tests/tt_metal/tt_metal/benchmark/padding/pad_rm.cpp))
- Add python `from_torch` benchmark to verify speed up in `from_torch` call ([tests/ttnn/benchmark/benchmark_from_torch.py](https://github.com/tenstorrent/tt-metal/blob/philei/speed-up-host-padding/tests/ttnn/benchmark/benchmark_from_torch.py))
- New python dependency for developers `pytest-benchmark==4.0.0` with BSD 2-Clause License shouldn't be a problem (Please let me know if I'm wrong), and I think other developers can add more ttnn benchmarks in the future

### Results
Note: Speed-up will most probably vary a lot depending on the CPU and RAM used in the system.
Achieved speed up is up to ~4.79x: 

C++ padding benchmark 
```
// old padding
------------------------------------------------------------------------------------------
Benchmark                                                Time             CPU   Iterations
------------------------------------------------------------------------------------------
BM_pad_rm_2d_last_dim_right/min_time:5.000             283 ms          283 ms           22
BM_pad_rm_2d_last_dim_left_right/min_time:5.000        283 ms          283 ms           25
BM_pad_rm_4d_last_dim_left_right/min_time:5.000        505 ms          505 ms           10

// new padding
------------------------------------------------------------------------------------------
Benchmark                                                Time             CPU   Iterations
------------------------------------------------------------------------------------------
BM_pad_rm_2d_last_dim_right/min_time:5.000            59.1 ms         59.1 ms          117
BM_pad_rm_2d_last_dim_left_right/min_time:5.000       59.6 ms         59.6 ms          117
BM_pad_rm_4d_last_dim_left_right/min_time:5.000        125 ms          125 ms           55
```
Python `from_torch` benchmark
```
// old padding
--------------------------------------------------------------------------------------------------------- benchmark: 3 tests ---------------------------------------------------------------------------------------------------------
Name (time in us)                                Min                     Max                    Mean              StdDev                  Median                   IQR            Outliers           OPS            Rounds  Iterations
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_benchmark_from_torch_zero_copy           6.9687 (1.0)            7.1614 (1.0)            7.0623 (1.0)        0.0631 (1.0)            7.0807 (1.0)          0.1167 (1.0)           4;0  141,597.8123 (1.0)          10         100
test_benchmark_from_torch_one_copy       71,635.8966 (>1000.0)   72,081.9220 (>1000.0)   71,800.7976 (>1000.0)  181.0103 (>1000.0)   71,739.7020 (>1000.0)    259.9362 (>1000.0)       1;0       13.9274 (0.00)          5          10
test_benchmark_from_torch_two_copy      357,099.3121 (>1000.0)  359,099.3495 (>1000.0)  358,294.4295 (>1000.0)  774.7156 (>1000.0)  358,435.1634 (>1000.0)  1,054.8201 (>1000.0)       2;0        2.7910 (0.00)          5          10
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------


// new padding
---------------------------------------------------------------------------------------------------------- benchmark: 3 tests ----------------------------------------------------------------------------------------------------------
Name (time in us)                                Min                     Max                    Mean                StdDev                  Median                   IQR            Outliers           OPS            Rounds  Iterations
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_benchmark_from_torch_zero_copy           6.9719 (1.0)            7.4110 (1.0)            7.0987 (1.0)          0.1366 (1.0)            7.0622 (1.0)          0.1696 (1.0)           1;0  140,870.8836 (1.0)          10         100
test_benchmark_from_torch_one_copy       69,002.4484 (>1000.0)   69,365.7713 (>1000.0)   69,194.4890 (>1000.0)    137.1384 (>1000.0)   69,178.9298 (>1000.0)    186.3429 (>1000.0)       2;0       14.4520 (0.00)          5          10
test_benchmark_from_torch_two_copy      120,715.6849 (>1000.0)  123,811.4174 (>1000.0)  122,418.0859 (>1000.0)  1,342.7287 (>1000.0)  122,176.9094 (>1000.0)  2,333.7123 (>1000.0)       2;0        8.1687 (0.00)          5          10
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes -- https://github.com/tenstorrent/tt-metal/actions/runs/14758338301 
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] **(For models and ops writers)** Full [new models tests](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) CI passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: ttnn/cpp/ttnn/tensor/tensor_impl.cpp [hunk 1]
```diff
@@ -1322,56 +1322,68 @@ Tensor pad(
...
+                if (coords[dim] < input_padded_shape[dim]) {
+                    break;
                 }
+                coords[dim] = 0;
             }
 
-            for (auto i = 0; i < pad_size[dim][1] * output_strides[dim]; i++) {
-                output_buffer[flat_output_index++] = pad_value_;
+            // Exit condition: all dimensions were processed
+            if (dim < 0) {
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/tensor/tensor_impl.cpp:1383]: Please explain how this won't spin forever in a corner case or add a counter to exit if there are too many iterations.

**Review by bbradelTT — APPROVED** [review]
Please test with a 0D tensor (not 0 volume).

File: ttnn/cpp/ttnn/tensor/tensor_impl.cpp (hunk 1 above)
Comment (philei-tt) [reply ttnn/cpp/ttnn/tensor/tensor_impl.cpp:1383]: Loop before that basically iterates over all coordinates in row major manner, and if we reached the maximum of first dim, that means we iterated over everything. I have rewritten that to make a bit more explicit. Let me know if you still need more clarity on that one

**Comment by philei-tt** [comment]
@tenstorrent/metalium-developers-infra @pgkeller @dmakoviichuk-tt ptal

File: ttnn/cpp/ttnn/tensor/tensor_impl.cpp [hunk 2]
```diff
@@ -1318,56 +1318,67 @@ Tensor pad(
...
-            output_strides.push_back(compute_stride(output_padded_shape, index));
+        // Calculate strides
+        ttnn::SmallVector<size_t> input_strides(rank);
+        ttnn::SmallVector<size_t> output_strides(rank);
+        input_strides[rank - 1] = 1;
+        output_strides[rank - 1] = 1;
+        for (int i = rank - 2; i >= 0; --i) {
+            input_strides[i] = input_strides[i + 1] * input_padded_shape[i + 1];
+            output_strides[i] = output_strides[i + 1] * output_padded_shape[i + 1];
         }
```
Comment (omilyutin-tt) [inline ttnn/cpp/ttnn/tensor/tensor_impl.cpp:1344]: Btw we have `compute_strides` function in `shape.hpp` to do this. Could be re-used?

File: tests/ttnn/benchmark/cpp/padding/pad_rm.cpp [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+#include <tt_stl/indestructible.hpp>
+#include <ttnn/tensor/storage.hpp>
+#include <tt-metalium/bfloat16.hpp>
+#include <ttnn/tensor/host_buffer/functions.hpp>
+#include <ttnn/tensor/host_buffer/host_buffer.hpp>
+#include "small_vector.hpp"
+
+namespace {
+template <typename T>
+ttnn::Tensor GenInputTensor(const ttnn::SmallVector<uint32_t>& shape) {
```
Comment (omilyutin-tt) [inline tests/ttnn/benchmark/cpp/padding/pad_rm.cpp:20]: A couple of minor things:
* I think all of test functions can be encapsulated in unnamed namespace, to enforce internal linkage
* `static` is implied if that is the case.

Comment (omilyutin-tt) [reply tests/ttnn/benchmark/cpp/padding/pad_rm.cpp:20]: Also `get_input_tensor`

**Review by omilyutin-tt — APPROVED** [review]

**Review by blozano-tt — APPROVED** [review]

File: tests/ttnn/benchmark/cpp/padding/pad_rm.cpp [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    }
+    state.SetComplexityN(N_padded);
+}
+
+BENCHMARK(BM_pad_rm_2d_last_dim_right)->Unit(benchmark::kMillisecond)->MinTime(5.0);
+BENCHMARK(BM_pad_rm_2d_last_dim_left_right)->Unit(benchmark::kMillisecond)->MinTime(5.0);
+BENCHMARK(BM_pad_rm_4d_last_dim_left_right)->Unit(benchmark::kMillisecond)->MinTime(5.0);
+BENCHMARK(BM_pad_rm_2d_scaling)->Unit(benchmark::kMillisecond)->RangeMultiplier(2)->Range(128, 8192)->Complexity();
+
+int main(int argc, char** argv) {
```
[the remark this answers is not in the document]
Comment (blozano-tt) [reply tests/ttnn/benchmark/cpp/padding/pad_rm.cpp:108]: Also curious why we have to implement `int main` with a test harness.

**Review by bkeith-TT — APPROVED** [review]
Bringing dependencies in with fixed versions is incurring technical debt, but is consistent with the status quo.

File: ttnn/cpp/ttnn/tensor/tensor_impl.cpp (hunk 2 above)
Comment (philei-tt) [reply ttnn/cpp/ttnn/tensor/tensor_impl.cpp:1344]: Indeed! I will try that

File: tests/ttnn/benchmark/cpp/padding/pad_rm.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (philei-tt) [reply tests/ttnn/benchmark/cpp/padding/pad_rm.cpp:108]: That's right, BENCHMARK_MAIN works here, I will change that in the next commit

[the remark this answers is not in the document]
Comment (philei-tt) [reply tests/ttnn/benchmark/cpp/padding/pad_rm.cpp]: Updated

File: tests/ttnn/benchmark/cpp/padding/pad_rm.cpp (hunk 3 above)
Comment (philei-tt) [reply tests/ttnn/benchmark/cpp/padding/pad_rm.cpp:20]: You're right, I moved everything under namespace {}

**Review by ayerofieiev-tt — APPROVED** [review]
